File: beacon-sdk-java/src/main/java/io/beacon/sdk/metrics/SdkMetrics.java
package io.beacon.sdk.metrics;

import java.util.concurrent.atomic.AtomicLong;

/**
 * SDK self-observability counters/gauges per spec/02 §3.
 *
 * <p>Six metrics in the spec: {@code records_enqueued}, {@code records_dropped},
 * {@code records_exported}, {@code export_failures}, {@code buffer_depth},
 * {@code fallback_writes}. M1.2 implements the three driven by the emit path;
 * M1.3 adds {@code batches_flushed} + {@code records_flushed} for the batch
 * flusher's observability; M1.4 wires {@code exported} + {@code export_failures}
 * + {@code fallback_writes} for the exporter/resilience layer.</p>
 */
public final class SdkMetrics {

    private final AtomicLong enqueued = new AtomicLong();
    private final AtomicLong dropped = new AtomicLong();
    private final AtomicLong bufferDepth = new AtomicLong();
    private final AtomicLong batchesFlushed = new AtomicLong();
    private final AtomicLong recordsFlushed = new AtomicLong();
    private final AtomicLong exported = new AtomicLong();
    private final AtomicLong exportFailures = new AtomicLong();
    private final AtomicLong fallbackWrites = new AtomicLong();

    // ---- M1.2 surface ---------------------------------------------------

    public void incEnqueued() { enqueued.incrementAndGet(); }
    public long enqueued() { return enqueued.get(); }

    public void incDropped() { dropped.incrementAndGet(); }
    public long dropped() { return dropped.get(); }

    public void setBufferDepth(int depth) { bufferDepth.set(depth); }
    public long bufferDepth() { return bufferDepth.get(); }

    // ---- M1.3 surface — batch flusher -----------------------------------

    public void incBatchesFlushed() { batchesFlushed.incrementAndGet(); }
    public long batchesFlushed() { return batchesFlushed.get(); }

    public void incRecordsFlushed(int n) { recordsFlushed.addAndGet(n); }
    public long recordsFlushed() { return recordsFlushed.get(); }

    // ---- M1.4 surface — exporter + resilience ---------------------------

    public void incExported(int n) { exported.addAndGet(n); }
    public long exported() { return exported.get(); }

    public void incExportFailure() { exportFailures.incrementAndGet(); }
    public long exportFailures() { return exportFailures.get(); }

    public void incFallbackWrite(int n) { fallbackWrites.addAndGet(n); }
    public long fallbackWrites() { return fallbackWrites.get(); }
}

File: beacon-sdk-python/src/beacon/metrics/sdk_metrics.py
"""SDK self-observability counters/gauges — Python idiom of Java ``SdkMetrics``.

Field-for-field intent parity with
``beacon-sdk-java/src/main/java/io/beacon/sdk/metrics/SdkMetrics.java`` per
spec/02 §3. The Java class uses ``AtomicLong`` per counter; Python has no atomic
long, so every mutation and read acquires a single ``threading.Lock``. This is
the deliberate Python idiom of Java ``AtomicLong``.

``itertools.count`` is NOT used: it is not safe for the read pattern (you cannot
read its current value without consuming it, and concurrent ``next()`` plus a
separate read would lose updates) — the lock-guarded plain ``int`` is correct
and the concurrent-increment unit test proves no lost updates.

Scope: M2.1 owns the three emit-path metrics (``records_enqueued``,
``records_dropped``, ``buffer_depth``). The remaining counters from spec/02 §3
fill in across later phases, mirroring the Java class's staged surface:
``records_exported`` + ``export_failures`` + ``fallback_writes`` (M2.3
exporter/resilience), ``batches_flushed`` + ``records_flushed`` (M2.2 flusher),
``redactor_timeouts`` (M2.5 redactor).
"""

from __future__ import annotations

import threading


class SdkMetrics:
    """Lock-guarded counters/gauge for the emit path (M2.1) + flusher path (M2.2).

    Every accessor and mutator takes ``self._lock`` so increments never
    interleave-and-lose under concurrent producers — the Python idiom of Java's
    ``AtomicLong``.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._enqueued = 0
        self._dropped = 0
        self._buffer_depth = 0
        self._batches_flushed = 0
        self._records_flushed = 0

    # ---- M2.1 surface — emit path --------------------------------------

    def inc_enqueued(self) -> None:
        """Increment ``records_enqueued`` by 1 (mirror Java ``incEnqueued``)."""
        with self._lock:
            self._enqueued += 1

    @property
    def enqueued(self) -> int:
        with self._lock:
            return self._enqueued

    def inc_dropped(self, n: int = 1) -> None:
        """Add ``n`` to ``records_dropped`` (default 1).

        ``DROP_OLDEST`` may pass a count for multi-eviction; ``DROP_NEWEST``
        uses the default. Mirror Java ``incDropped`` (with an additive variant).
        """
        with self._lock:
            self._dropped += n

    @property
    def dropped(self) -> int:
        with self._lock:
            return self._dropped

    def set_buffer_depth(self, depth: int) -> None:
        """Set the ``buffer_depth`` gauge (set, not add — mirror Java ``setBufferDepth``)."""
        with self._lock:
            self._buffer_depth = depth

    @property
    def buffer_depth(self) -> int:
        with self._lock:
            return self._buffer_depth

    # ---- M2.2 surface — flusher path -----------------------------------

    def inc_batches_flushed(self) -> None:
        """Increment ``batches_flushed`` by 1 (mirror Java ``incBatchesFlushed``)."""
        with self._lock:
            self._batches_flushed += 1

    @property
    def batches_flushed(self) -> int:
        with self._lock:
            return self._batches_flushed

    def inc_records_flushed(self, n: int) -> None:
        """Add ``n`` to ``records_flushed`` (mirror Java ``incRecordsFlushed(int)``).

        The flusher passes the flushed batch size on each successful flush.
        """
        with self._lock:
            self._records_flushed += n

    @property
    def records_flushed(self) -> int:
        with self._lock:
            return self._records_flushed
